kroxylicious/src/main/java/io/kroxylicious/proxy/config/ServiceBasedPluginFactoryRegistry.java

```diff
@@ -0,0 +1,126 @@
+                Stream.of(providerType.getName(), providerType.getSimpleName()).forEach(name2 -> nameToProviders.compute(name2, (k2, v) -> {
+                    if (v == null) {
+                        v = new HashSet<>();
+                    }
+                    v.add(providerAndConfigType);
+                    return v;
+                }));
+            }
+        });
+        var bySingleton = nameToProviders.entrySet().stream().collect(
+                Collectors.partitioningBy(e -> e.getValue().size() == 1));
+        if (LOGGER.isWarnEnabled()) {
+            for (Map.Entry<String, Set<ProviderAndConfigType>> ambiguousInstanceNameToProviders : bySingleton.get(false)) {
+                LOGGER.warn("'{}' would be an ambiguous reference to a {} provider. "
+                        + "It could refer to any of {}"
+                        + " so to avoid ambiguous behaviour those fully qualified names must be used",
+                        ambiguousInstanceNameToProviders.getKey(),
+                        pluginInterface.getSimpleName(),
+                        ambiguousInstanceNameToProviders.getValue().stream().map(p -> p.provider().type().getName()).collect(Collectors.joining(", ")));
+            }
+        }
+        return bySingleton.get(true).stream().collect(Collectors.toMap(
+                Map.Entry::getKey,
+                e -> e.getValue().iterator().next()));
+    }
+
+    @Override
+    public <P> @NonNull PluginFactory<P> pluginFactory(@NonNull Class<P> pluginClass) {
+        var nameToProvider = load(pluginClass);
+        return new PluginFactory<>() {
+            @Override
+            public @NonNull P pluginInstance(@NonNull String instanceName) {
+                if (Objects.requireNonNull(instanceName).isEmpty()) {
+                    throw new IllegalArgumentException();
+                }
+                var provider = nameToProvider.get(instanceName);
+                if (provider != null) {
+                    Class<?> type = provider.provider().type();
+                    if (type.isAnnotationPresent(Deprecated.class)) {
+                        LOGGER.warn("{} plugin with id {} is deprecated",
+                                pluginClass.getName(),
+                                instanceName);
+                    }
+                    return pluginClass.cast(provider.provider().get());
+                }
+                throw unknownPluginInstanceException(instanceName);
+            }
+
+            private UnknownPluginInstanceException unknownPluginInstanceException(String name) {
+                return new UnknownPluginInstanceException("Unknown " + pluginClass.getName() + " plugin instance for name '" + name + "'. "
+                        + "Known plugin instances are " + nameToProvider.keySet() + ". "
+                        + "Plugins must be loadable by java.util.ServiceLoader.");
```

Comment:
we emitting a warning when the developer misses the annotation, but they're likely not to notice that.   How about?

```suggestion
                        + "Plugins must be loadable by java.util.ServiceLoader and annotated with @PluginConfigType");
```